system/lib/wasmfs/file.cpp

```diff
@@ -32,13 +32,59 @@ void DataFile::Handle::preloadFromJS(int index) {
 // Directory
 //
 std::shared_ptr<File> Directory::Handle::getEntry(std::string pathName) {
-  auto it = getDir()->entries.find(pathName);
-  if (it == getDir()->entries.end()) {
-    return nullptr;
-  } else {
-    return it->second;
+  for (const auto& entry : getDir()->entries) {
+    if (entry.name == pathName) {
+      return entry.file;
+    }
+  }
+  return nullptr;
+}
+
+void Directory::Handle::setEntry(std::string pathName,
+                                 std::shared_ptr<File> inserted) {
+  // Hold the lock over both functions to cover the case in which two
+  // directories attempt to add the file.
+  auto lockedInserted = inserted->locked();
+  // Simultaneously, set the parent of the inserted node to be this Dir.
+  // inserted must be locked because we have to go through Handle.
+  // TODO: When rename is implemented, ensure that the source directory has
+  // been removed as a parent.
+  // https://github.com/emscripten-core/emscripten/pull/15410#discussion_r742171264
+  assert(!lockedInserted.getParent());
+  lockedInserted.setParent(file);
+
+  // During testing, this will check that an existing file associated with
+  // pathName does not exist. For rename, the existing file must be unlinked
+  // first.
+  assert(!getEntry(pathName));
+  getDir()->entries.push_back({pathName, inserted});
+}
+
+void Directory::Handle::unlinkEntry(std::string pathName) {
+  // The file lock must be held for both operations. Removing the child file
+  // from the parent's entries and removing the parent pointer from the
+  // child should be atomic. The state should not be mutated in between.
+  auto unlinked = getEntry(pathName)->locked();
+  unlinked.setParent({});
+
+  for (auto it = getDir()->entries.begin(); it != getDir()->entries.end();
+       it++) {
+    if (it->name == pathName) {
+      getDir()->entries.erase(it);
+      return;
+    }
```

Comment:
This loop could be replaced with a `std::remove` + `std::erase` or `std::find` + `vector::erase`.